[Inductor][Grouped Gemm] Add Blackwell CuTeDSL Kernel

NikhilAPatel opened the pull request:
Make sure you're on cutlass 4.2.0+


Test Plan:
Tritonbench(oss):
`clear; CUDA_VISIBLE_DEVICES=2 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 python run.py --op grouped_gemm --only aten_grouped_mm,preprocessed_pt2_triton_grouped_mm --precision bf16  --num-inputs 1 --metrics tflops,accuracy`

Unit Tests(oss):
`clear; python test/inductor/test_cutedsl_grouped_mm.py`

Differential Revision: D82010227


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

alexsamardzic reviewed (approved):
This is awesome!  Looking forward to see the performance numbers.

EDIT:  Grabbed a B100 over a short period of time, some further comments:

1. Would be good to update `test_grouped_gemm_compiled` from `test/test_matmul_cuda.py` with this kernel as an option, as this test has some nasty offsets, and on a quick try it seems the kernel won't pass (2d/3d "op", both A and B "row-major" and max-autotune turned on).  Or, even better: to merge mentioned test and tests from `test/inductor/test_cutedsl_grouped_mm.py` into single place, to test with union of shapes/offsets, and to test both back-ends.
2. Tried to do a quick benchmark with the script pasted below, but latencies reported by this script for CuTe DSL are very large - apparently, proper compilation/tuning are yet to be implemented?   (Triton kernel speedups are low on Blackwell too, 0.1 and below for larger M, obviously for Blackwell and later, kernels will have to be written in a lower-level language.)
3. @ngimel: Not related to this PR, but I think "sm100a" should be added in the list [here](https://github.com/pytorch/pytorch/blob/a2601630cd9850ec68da8456d9f9584da9c0b0e8/cmake/Codegen.cmake#L132)?

<details>
<summary>Benchmarking script</summary>

```Python
import pandas as pd

from tqdm import tqdm

import torch
from triton.testing import do_bench


device = "cuda"
dtype_AB = torch.bfloat16
dtype_offset = torch.int32
dtype_C = torch.bfloat16
group_size = 4
strided = True

f_ref = torch._grouped_mm
f = torch.compile(
    f_ref,
    options={
        "max_autotune": True,
        "max_autotune_gemm_backends": "CUTE",
        "autotune_fallback_to_aten": False,
    },
)


def benchmark_microseconds(f, *args, **kwargs):
    return do_bench(lambda: f(*args, **kwargs), return_mode="median") * 1e3


def get_problem(M, N, K):
    A = torch.randn(M * group_size, K, device=device).to(dtype_AB)
    B = torch.randn(group_size, N, K, device=device).to(dtype_AB)
    offs = torch.arange(M, group_size * M + 1, M, device=device, dtype=dtype_offset)

    return A, B, offs


def benchmark(m: int, k: int, n: int):
    for i in range(5):
        A, B, offs = get_problem(m, n, k)
        C = f(A, B.transpose(-2, -1), offs, out_dtype=dtype_C)
        C_ref = f(A, B.transpose(-2, -1), offs, out_dtype=dtype_C)
        torch.testing.assert_close(C_ref, C)

    A, B, offs = get_problem(m, n, k)

    eager_time = benchmark_microseconds(
        f_ref,
        A,
        B.transpose(-2, -1),
        offs,
        out_dtype=dtype_C,
    )
    compiled_time = benchmark_microseconds(
        f,
        A,
        B.transpose(-2, -1),
        offs,
        out_dtype=dtype_C,
    )

    return {
        "m": m,
        "n": n,
        "k": k,
        "Eager (CUTLASS) latency (us)": eager_time,
        "Compiled (CuTe) latency (us)": compiled_time,
        "Compiled speedup (d/s)": eager_time / compiled_time,
    }


if __name__ == "__main__":
    k_vals = (4096, 8192)
    n_vals = (4096, 8192)
    results = []
    i_range = range(8, 14)
    for m in tqdm([1 << i for i in i_range]):
        for n, k in zip(n_vals, k_vals):
            results.append(benchmark(m, k, n))

        df = pd.DataFrame(results)
        df.to_csv(f"grouped_mm_2d_3d_time_results_cute.csv", index=False)
        print(df.to_markdown(index=False))
```
</details>

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,287 @@
+            stream,
+        )
+
+        cache["{{kernel_name}}"] = (build_group_ptrs_jit_executor, group_gemm_jit_executor, total_num_clusters, tensormap_shape)
```
for posterity, talked today, currently since we splat out a new file and thus a new cache = {} for all the constexprs we dont need to actually add any of these to the cache key. We also essentially have 1 cache per kernel name so don't need to include.

Empirically the kernels does specialize on strides and will get bad results we dont recompile, for the preproc kernel. 

We are going to only enable for static shapes and then cache all the things so expect lots of recomptue at first and then will follow up w/ more robust dynamic shape caching

ngimel replied:
The dimensions should be reasonably static though? Except possibly for M, though M with dropped tokens should be static too? Also can we have more detailed comments about CuTE caching, currently it's behavior is pretty opaque

eellison commented on test/inductor/test_cutedsl_grouped_mm.py:
```diff
@@ -0,0 +1,146 @@
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+)
```
Do we need to add this test to b200 test files ?

NikhilAPatel replied:
Yeah we probably do I'll add it.

eellison commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,329 @@
+
+    cache_key = get_cache_key(input_a, input_b, out_ptr0)
+
+    if cache_key not in cache:
```
how long does it take to compile ? would we want to interleave the compilation in the future?

ngimel replied:
Is it for all the autotuning instantiations or just one?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -1870,6 +1870,77 @@ def use_triton_blackwell_tma_template(
+        return False
+
+    # Checks for 16B ptr and stride alignment
+    if not can_use_tma(mat_a, mat_b, output_layout=layout):
```
i guess aligned is already constrained here

eellison commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,329 @@
+    ptrs_t = torch.empty((G, 3), device=dev, dtype=torch.int64)
+    probs_t = torch.empty((G, 4), device=dev, dtype=torch.int32)
+    strides_t = torch.empty((G, 3, 2), device=dev, dtype=torch.int32)
+    ptrs = from_dlpack(ptrs_t)
```
why do we need `from_dlpack ` , just wondering?

NikhilAPatel replied:
This is to convert from PyTorch tensors to CuTe tensors. You can see more information here: https://docs.nvidia.com/cutlass/media/docs/pythonDSL/cute_dsl_general/framework_integration.html

eellison reviewed (commented):
cool ! deferring to @drisspg and others for final review

NikhilAPatel commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,329 @@
+    return (sm_count, max_active_clusters)
+
+
+def get_cache_key(input_a: torch.Tensor, input_b: torch.Tensor, out_ptr0: torch.Tensor):
```
This kernel assumes 16B alignment. We verify this at compilation time. Do you know if this guarantees that tensors will be aligned at runtime as well?

drisspg replied:
My quick pass makes me think that the copy_ alignment path will also just `just work` for inputs to a graph w/ cutedsl templates but feel free to correct me here @eellison

drisspg commented on .ci/pytorch/test.sh:
```diff
@@ -337,7 +337,7 @@ test_python() {
 test_python_smoke() {
   # Smoke tests for H100/B200
-  time python test/run_test.py --include test_matmul_cuda test_scaled_matmul_cuda inductor/test_fp8 inductor/test_max_autotune $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
+  time python test/run_test.py --include test_matmul_cuda test_scaled_matmul_cuda inductor/test_fp8 inductor/test_max_autotune inductor/test_cutedsl_grouped_mm $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
```
One note is that this actually wont testing anything because we won't have cutedsl install I will fix when I figure out packaging

drisspg commented on test/inductor/test_cutedsl_grouped_mm.py:
```diff
@@ -0,0 +1,146 @@
+        dtype: torch.dtype,
+    ) -> tuple[Tensor, Tensor, Tensor]:
+        # --- Random, tile-aligned M sizes ---
+        M_sizes = (
```
Also is 64 a requirement or is 16 bytes?

drisspg commented on test/inductor/test_cutedsl_grouped_mm.py:
```diff
@@ -0,0 +1,146 @@
+
+        self.assertEqual(c_eager.dtype, dtype)
+        self.assertEqual(c_compiled.dtype, dtype)
+        torch.testing.assert_close(c_eager, c_compiled, rtol=1e-2, atol=0.05)
```
how does this comapare to other compiled v eager numerics for grouped_mm?

NikhilAPatel replied:
Removed the manual overrides of rtol and atol and switched to the default for bf16 used by `torch.testing.assert_close` (rtol=1.6e-2, atol=1e-5). These are the same defaults used by `test_grouped_gemm_compiled` in test_matmul_cuda. We still pass, so numerics seem good.

drisspg commented on test/inductor/test_cutedsl_grouped_mm.py:
```diff
@@ -0,0 +1,146 @@
+        if layout_A == "offset":
+            # allocate bigger buffer than needed, use nonzero storage offset
+            storage = torch.randn(sum_M * K + 512, device=device, dtype=dtype)
+            offset = 128  # skip first 128 elements
```
if you choose a more unfriendly number what happends?

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,329 @@
+# redundant recompiles even when shape/stride/dtype cache misses force kernel
+# regeneration. A follow-up study will investigate the root cause.
+
+cache = {}
```
yup to both :)

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped.py.jinja:
```diff
@@ -0,0 +1,329 @@
+            total_num_clusters = 0
+            for m, n, _, _ in problem_sizes_mnkl:
+                num_clusters_mn = tuple(
+                    (x + y - 1) // y for x, y in zip((m, n), cluster_tile_shape_mn)
```
nit: ceildiv

drisspg reviewed (approved):
Small nits and 1 comment, overall looks good

We should track somewhere, maybe open an issue? on the follow up tasks

albanD commented on test/test_public_bindings.py:
```diff
@@ -363,6 +363,7 @@ def onerror(modname):
             "torch._inductor.codegen.cpp_micro_gemm",
             "torch._inductor.codegen.cpp_template_kernel",
             "torch._inductor.runtime.triton_helpers",
+            "torch._inductor.template_heuristics.cutedsl",
```
Any way we could make this module actually importable? It's weird to ship in our package something that looks like a submodule but should not be imported

NikhilAPatel commented on torch/_inductor/config.py:
```diff
@@ -534,6 +534,10 @@ def prologue_fusion_enabled() -> bool:
     "TORCHINDUCTOR_MAX_AUTOTUNE_FLEX_SEARCH_SPACE", "DEFAULT"
 ).upper()  # type: ignore[assignment]
 
+cutedsl_enable_autotuning: bool = (
```
Right now autotuning is gated behind a config variable – CUTEDSL_ENABLE_AUTOTUNING – since I find that turning it on results in a flurry of call cuLaunchKernel/cuLaunchKernelEx failed, got error code: 1 errors. It still seems, however, that autotuning finishes successfully, so I’m still digging into why this occurs and what exactly the implications are. Soon I hope this can be enabled by default.

mlazos commented on test/inductor/test_cutedsl_grouped_mm.py:
```diff
@@ -0,0 +1,148 @@
+            {
+                "max_autotune": True,
+                "max_autotune_gemm_backends": "CUTEDSL",
+                "test_configs.autotune_choice_name_regex": "cutedsl",
```
hmm are there other choices that get added when you have only the single backend?

mlazos commented on torch/_inductor/template_heuristics/cutedsl.py:
```diff
@@ -0,0 +1,293 @@
+            USE_2_CTA=False,
+            TENSORMAP_UPDATE_MODE=TensorMapUpdateMode.GMEM,
+        ),
+        CuTeGemmConfig(
```
nit: for cleaning up the code a bit can you make these a list of tuples and then apply the constructor to the unpacked tuples? e.g. ([CuTeGemmConfig(*args) for args in configs]) and if there's a pattern can you autogenerate?

alexsamardzic commented:
In the meantime I came up with [this](https://github.com/pytorch/pytorch/pull/166063), while experimenting with Gluon; I've included this CuTe DSL kernel into benchmarking.  The CuTe DSL kernel is close to CUTLASS `torch._grouped_mm` with performance, but it seems there is a whole load of elementwise operators called before the MM kernel, and after `build_group_ptrs_from_bases` kernel, that seem to take almost the same time as the MM kernel itself.   Attached is a screenshot of profiler trace in Perfetto, maybe worth checking.

<img width="3606" height="550" alt="perfetto" src="https://github.com/user-attachments/assets/ae2bce1e-bf9b-4d63-be0c-cc80521a6b12" />

ngimel commented:
@NikhilAPatel did you have a chance to look into @alexsamardzic comment? What is causing the elementwise kernels before grouped_mm?

mlazos commented:
Also did you figure out the slowdown @NikhilAPatel ? The one we talked about in our 1:1, not the elementwise kernels.

NikhilAPatel committed:
[Inductor][Grouped Gemm] Add Blackwell CuTeDSL Kernel (#165036)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/165036

Test Plan:
Inductor test (fbcode):
`INDUCTOR_TEST_DISABLE_FRESH_CACHE=1 TORCHINDUCTOR_CACHE_DIR=~/cutetest buck2 run mode/opt //caffe2/test/inductor:cutedsl_grouped_mm -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 -m "ovr_config//third-party/pypi/nvidia-cutlass-dsl/constraints:4.2.1"`



Tritonbench (fbcode):
`clear; CUDA_VISIBLE_DEVICES=7 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 buck2 run mode/opt //pytorch/tritonbench:run -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 -m "ovr_config//third-party/pypi/nvidia-cutlass-dsl/constraints:4.2.1" -- --op grouped_gemm --only aten_grouped_mm,preprocessed_pt2_cute_grouped_mm --precision bf16  --num-inputs 1 --metrics tflops,accuracy`


Tritonbench(oss):
`clear; CUDA_VISIBLE_DEVICES=2 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 python run.py --op grouped_gemm --only aten_grouped_mm,preprocessed_pt2_triton_grouped_mm --precision bf16  --num-inputs 1 --metrics tflops,accuracy`


Unit Tests(oss):
`clear; python test/inductor/test_cutedsl_grouped_mm.py`

Reviewed By: drisspg

Differential Revision: D82010227

atalman commented:
@pytorchmergebot revert -c nosignal -m "regressed vllm signal: [GH job link](https://github.com/pytorch/pytorch/actions/runs/19059329909/job/54439919668) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/0e1a88904f4a5e30634b196678b56e1d6ec074f5)"
